File: packages/binding/src/bindings/go/request.go
package main

import "syscall/js"
import "encoding/json"
%%IMPORTS%%

type t_request func(Request) any
type t_response func(js.Value, []js.Value) any
type Object map[string]any
type Array []any

type Dispatcher struct {
  Get func(string) any
  All func() map[string]any
  %%DISPATCH_STRUCT%%
}

type URL struct {
  Href string
  Origin string
  Protocol string
  Username string
  Password string
  Host string
  Hostname string
  Port string
  Pathname string
  Search string
  SearchParams map[string]any
  Hash string
}

type Request struct {
  Url URL
  Body Dispatcher
  Path Dispatcher
  Query Dispatcher
  Cookies Dispatcher
  Headers Dispatcher
  %%REQUEST_STRUCT%%
}

func make_url(request js.Value) URL {
  url := request.Get("url");
  search_params := make(map[string]any);
  json.Unmarshal([]byte(request.Get("search_params").String()), &search_params);

  return URL{
    url.Get("href").String(),
    url.Get("origin").String(),
    url.Get("protocol").String(),
    url.Get("username").String(),
    url.Get("password").String(),
    url.Get("host").String(),
    url.Get("hostname").String(),
    url.Get("port").String(),
    url.Get("pathname").String(),
    url.Get("search").String(),
    search_params,
    url.Get("hash").String(),
  };
}

func make_dispatcher(key string, request js.Value) Dispatcher {
  properties := make(map[string]any);
  value := request.Get(key)
  json.Unmarshal([]byte(value.Get("properties").String()), &properties);

  return Dispatcher{
    func(property string) any {
      switch properties[property].(type) {
        case string:
          return properties[property].(string);
        default:
          return nil;
        }
    },
    func() map[string]any {
      return properties;
    },
    %%DISPATCH_MAKE%%
  };
}

func make_request(route t_request, request js.Value) any {
  go_request := Request{
    make_url(request),
    make_dispatcher("body", request),
    make_dispatcher("path", request),
    make_dispatcher("query", request),
    make_dispatcher("cookies", request),
    make_dispatcher("headers", request),
    %%REQUEST_MAKE%%
  };

  response := route(go_request);
  switch response.(type) {
    case js.Func:
      return response;
    default:
      marshalled, _ := json.Marshal(response);
      return string(marshalled);
    }
}

func make_empty(route func() any) any {
  response := route();
  switch response.(type) {
    case js.Func:
      return response;
    default:
      marshalled, _ := json.Marshal(response);
      return string(marshalled);
    }
}

File: packages/binding/src/bindings/ruby/request.rb
require 'json'

class Primate
  def self.view(name, props = {}, options = {})
    {:__PRMT__ => "view", :name => name, :props => props, :options => options}
  end

  def self.redirect(location, options = {})
    {:__PRMT__ => "redirect", :location => location, :options => options}
  end

  def self.error(body, options = {})
    {:__PRMT__ => "error", :body => body, :options => options}
  end
end

class Dispatcher
  def initialize(dispatcher)
    @dispatcher = dispatcher
    @json = JSON.parse(dispatcher.call("toString").to_s)
  end

  def get(name)
    @dispatcher.call("get", name).to_s
  end

  def json() 
    @json 
  end

  %%DISPATCH_DEFS%%
end

class URL
  def initialize(url)
    @href = url["href"].to_s
    @origin = url["origin"].to_s
    @protocol = url["protocol"].to_s
    @username = url["username"].to_s
    @password = url["password"].to_s
    @host = url["host"].to_s
    @hostname = url["hostname"].to_s
    @port = url["port"].to_s
    @pathname = url["pathname"].to_s
    @search = url["search"].to_s
    @hash = url["hash"].to_s
  end

  def href
    @href
  end

  def origin
    @origin
  end

  def protocol
    @protocol
  end

  def username
    @username
  end

  def password
    @password
  end

  def host
    @host
  end

  def hostname
    @hostname
  end

  def port
    @port
  end

  def pathname
    @pathname
  end

  def search
    @search
  end

  def hash
    @hash
  end
end

class Request
  def initialize(request)
    @url = URL.new(request["url"])
    @body = request["body"]
    @path = Dispatcher.new(request["path"])
    @query = Dispatcher.new(request["query"])
    @headers = Dispatcher.new(request["headers"])
    @cookies = Dispatcher.new(request["cookies"])
  end

  def url
    @url
  end

  def body
    @body
  end

  def path
    @path
  end

  def query
    @query
  end

  def headers
    @headers
  end

  def cookies
    @cookies
  end
end
